[REVIEW] Add a Stream class that wraps CuPy/Numba/CudaStream

shwina opened the pull request:
Replaces `CudaStreamView` with a slightly more general `Stream` class. A `Stream` can be created in a few different ways:

```python
>>> from rmm._cuda.stream import Stream
>>> import cupy
>>> from numba import cuda
>>> Stream()  # new CUDA stream
>>> Stream(cupy.cuda.Stream())  # from CuPy stream
>>> Stream(cuda.stream())  # from Numba stream
```

A `Stream` is composed of a `cudaStream_t` (`._ptr`), and optionally, a reference to a Python object that owns the stream (`._owner`). The `._ptr` attribute is only accessible from Cython.

When creating a `Stream` from scratch, the owner is a `rmm._lib.cuda_stream.CudaStream` object (a wrapper around the owning C++ class `rmm::cuda_stream`).

```python
>>> s = Stream()
>>> s._owner
<rmm._lib.cuda_stream.CudaStream object at 0x7ff540519510>
```

When creating a `Stream` from an existing stream object (such as a CuPy or Numba stream), the owner is that CuPy or Numba stream:

```python
>>> from numba import cuda
>>> s = Stream(cuda.stream())
>>> s._owner
<CUDA stream 93966239561424 on <CUDA context c_void_p(93966231976304) of device 0>>
```

Finally, the module constants `DEFAULT_STREAM`, `LEGACY_DEFAULT_STREAM` and `PER_THREAD_DEFAULT_STREAM` are `Stream` objects that refer to the default, legacy default, and per-thread default CUDA streams respectively:

```python
>>> from rmm._cuda.stream import DEFAULT_STREAM, LEGACY_DEFAULT_STREAM, PER_THREAD_DEFAULT_STREAM
```

cc: @kkraus14 @harrism

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

shwina commented:
rerun tests

jakirkham commented:
cc @pentschev

jakirkham commented:
cc @gmarkall

kkraus14 commented:
Lets push this to 0.18.

kkraus14 commented on python/rmm/_cuda/stream.pxd:
```diff
@@ -0,0 +1,28 @@
+        object _owner
+
+    cdef cuda_stream_view view(self) nogil except *
+    cpdef bool is_default(self) except *
+    cpdef void synchronize(self) except *
```
Can both of these be `nogil`?

shwina replied:
Not if they're `cpdef` :(

kkraus14 replied:
I could have sworn you can still mark cpdef functions as `nogil`

kkraus14 commented on python/rmm/_cuda/stream.pyx:
```diff
@@ -0,0 +1,90 @@
+    cudaStreamPerThread,
+)
+
+import cupy
+from numba import cuda
```
These should be optional dependencies

shwina replied:
CuPy is currently not a dependency, but numba is.

kkraus14 commented on python/rmm/_cuda/stream.pyx:
```diff
@@ -0,0 +1,90 @@
+        """
+        Synchronize the CUDA stream
+        """
+        self.view().synchronize()
```
We should explicitly release the GIL before synchronizing

kkraus14 commented on python/rmm/_lib/cuda_stream.pxd:
```diff
@@ -0,0 +1,34 @@
+
+cdef class CudaStream:
+    cdef unique_ptr[cuda_stream] c_obj
+    cpdef bool is_valid(self) except *
```
Can this be nogil?

shwina replied:
We'd need a `c_is_valid`. Is that OK?

kkraus14 replied:
I really dislike the `c_*` pattern for `cdef` apis. Do we have any alternatives?

shwina replied:
Yup, as discussed, I made CudaStream `@final`

kkraus14 commented on python/rmm/_cuda/stream.pyx:
```diff
@@ -0,0 +1,90 @@
+cdef class Stream:
+    def __init__(self, obj=None):
+        """
+        BA Stream represents a CUDA stream.
```
What is `BA Stream`?

shwina replied:
Intentionally introduced typo when playing with pre-commit and forgot :)

kkraus14 commented on python/rmm/_lib/cuda_stream.pyx:
```diff
@@ -0,0 +1,30 @@
+
+
+cdef class CudaStream:
+    """
+    Wraps rmm::cuda_stream (an owning class).
+    """
```
A more comprehensive docstring similar to `rmm.DeviceBuffer` would be nice here

shwina replied:
I'm not sure what to say about `CudaStream` other than it's a RAII wrapper around a CUDA stream. Ideally, no one should interact with `CudaStream` (just `Stream`). Anything specific you'd like to see here?

kkraus14 replied:
Maybe just add an example to the docstring?

harrism commented:
@shwina It bothers me that `Stream` can either be owning or non-owning of the underlying resource (the actual CUDA stream).  

Also, one of the main goals of the C++ work on `cuda_stream` and `cuda_stream_view` was to get away from integer stream handles as much as possible. Integer handles like `0` lead to ambiguity and therefore bugs. I think we should always refer to the default stream explicitly by a name rather than a zero. Likewise for per-thread default stream: by a name rather than a two.

raydouglass commented:
rerun tests

dillon-cullinan commented:
rerun tests

shwina commented:
@harrism 

> It bothers me that Stream can either be owning or non-owning of the underlying resource (the actual CUDA stream).

I'm not sure I understand this concern correctly. When would the a `Stream` object be considered "owning" v/s "non-owning"? It's true that a `Stream` object that is constructed from an integer handle is "non-owning", in that, it doesn't store a reference to a Python object that owns the underlying CUDA stream. `Stream` objects constructed in any other way are "owning" as they keep a reference to the Numba/CuPy/CudaStream object that owns the CUDA stream.

Please read on though, as I've removed the ability to construct from integer handles.

> Also, one of the main goals of the C++ work on cuda_stream and cuda_stream_view was to get away from integer stream handles as much as possible. Integer handles like 0 lead to ambiguity and therefore bugs.

Agreed. I've made the `_ptr` attribute a `cudaStream_t` instead. I've also removed the ability to construct `Streams` from integer handles. In its place, I've added the ability to construct them from `cudaStream_t`. Of course, the latter can only be used from Cython. The `DEFAULT_STREAM`, `LEGACY_STREAM` and `PER_THREAD_DEFAULT_STREAM` objects are now constructed directly from the `cudaStream_t` instances `cudaStreamDefault`, `cudaStreamLegacy` and `cudaStreamPerThread`. These three objects in particular are still "non-owning", in that they don't have any resources underneath them. Should they? Or can we assume that `cudaStreamDefault`, `cudaStreamLegacy` and `cudaStreamPerThread` are always valid?

kkraus14 commented on python/rmm/_cuda/stream.pxd:
```diff
@@ -0,0 +1,32 @@
+from rmm._lib.lib cimport cudaStream_t
+
+cdef class Stream:
+    cdef cudaStream_t _ptr
```
instead of this being `_ptr` should we make this something like `_cudastream_t`?

shwina replied:
Yes but here are some alternatives for your perusal: 

`_cuda_stream`
`_cuda_stream_t`
`_stream`
`c_obj`

kkraus14 replied:
I think `_stream` or `c_obj` would be a bit ambiguous. Fine with any of the others though 😄

harrism commented:
> > It bothers me that Stream can either be owning or non-owning of the underlying resource (the actual CUDA stream).
> 
> I'm not sure I understand this concern correctly. When would the a `Stream` object be considered "owning" v/s "non-owning"? It's true that a `Stream` object that is constructed from an integer handle is "non-owning", in that, it doesn't store a reference to a Python object that owns the underlying CUDA stream. `Stream` objects constructed in any other way are "owning" as they keep a reference to the Numba/CuPy/CudaStream object that owns the CUDA stream.

Owning classes and non-owning "view" classes are an important concept in libcudf (and in C++): https://github.com/rapidsai/cudf/blob/branch-0.17/cpp/docs/TRANSITIONGUIDE.md#views-and-ownership

Whenever you have an allocatable resource, it should be owned by an object. Streams are no different than memory in this regard. But CUDA streams are different because there are some streams you allocate (`cudaStreamCreate()`), and some that are built in (the null stream / `cudaStreamLegacy`, `cudaStreamPerThread`). So in RMM C++ we decided to differentiate between owning stream objects (`cuda_stream`) and non-owning stream objects (`cuda_stream_view`). It's the latter that is always used to pass a stream to a function that schedules work on the stream. The former is used only for lifetime management of non-default streams. This way there is no ambiguity about RAII. Views never manage lifetime. So you can only have a `cuda_stream_view` representing a default stream, and a `cuda_stream` only ever represents a non-default stream.

This PR conflates owned and non-owned streams. My main concern is that this is just different from what we do on the C++ side, and consistency is nice. But if conflating ownership with non-ownership is more Pythonic, I guess I can be convinced.

harrism commented:
BTW, CUDA made the amusing  choice of creating the `cudaStreamPerThread` and `cudaStreamLegacy` stream handle identifiers (which are defined as `(cudaStream_t)0x2` and `(cudaStream_t)0x1` respectively), but also defining something called `cudaStreamDefault`  which is NOT  a stream handle identifier. `cudaStreamDefault` is an option to `cudaStreamCreateWithFlags`! Notice that it  is defined as `0`, not `(cudaStream_t)0`. Therefore I have removed all usage of it in RMM as a stream identifier, and you should not use it as one in Cython code.

kkraus14 commented:
> The present PR conflates owned and non-owned streams.

I disagree. In this PR we have the `CudaStream` and `Stream` classes. `CudaStream` is an owning class, where once an instance's reference count goes to 0 a resource is freed under the hood. `Stream` on the other hand is a non-owning class, but holds references to owning classes (`CudaStream`, CuPy Stream, Numba Stream, etc.). In Python, essentially everything is a `shared_ptr` so you can think of our `Stream` class as similar to `cuda_stream_view` but it guarantees the lifetime of the underlying `cuda_stream`.

> My main concern is that this is just different from what we do on the C++ side, and consistency is nice. But if conflating ownership with non-ownership is more Pythonic, I guess I can be convinced.

The Pythonic was is to use Python reference counting which is what we're doing 😄.

harrism commented:
> I disagree.

No, you agree. I simply misunderstood.

shwina commented:
> cudaStreamDefault is an option to cudaStreamCreateWithFlags!

Wow, I definitely missed this. Thanks! So does the default constructor of `rmm::cuda_stream_view()` then provide a view into `(cudaStream_t)0x0`?

shwina commented on python/rmm/_cuda/stream.pyx:
```diff
@@ -0,0 +1,120 @@
+        self._cuda_stream, self._owner = stream._cuda_stream, stream._owner
+
+
+DEFAULT_STREAM = Stream._from_cudaStream_t(<cudaStream_t>0)
```
@harrism I replaced the use of `cudaStreamDefault` with `0` here. Is this appropriate?

harrism replied:
I would prefer you use `rmm::cuda_stream_default()`, which basically returns a default-constructed `cuda_stream_view`.

This way it's explicit, and if we decide environment-variable override of the semantics (so that it effectively returns 2 instead of 0) it will just work.

shwina replied:
We can't construct a `Stream` from a `rmm::cuda_stream_view`. We _can_ construct a `stream` from any of:

1. An `rmm::cuda_stream` (an owning class)
2. A CuPy stream (an owning class)
3. A Numba stream (an owning class)
4. A `cudaStream_t` (a non-owning class, but we only use this for `cudaStreamLegacy` and `cudaStreamDefault`)

Initially, I thought that it didn't make sense to construct Streams from `rmm::cuda_stream_view` objects because they are non-owning. But I guess makes sense to do so, even if only for the special case of `cudaStreamDefault`.

CUDA streams are weird :(

harrism replied:
You don't have to add construction from `rmm::cuda_stream_view()` to get rid of the explicit zero. Just call `value()`
https://github.com/rapidsai/rmm/blob/81881bf86c90c5370d4e917ba2ec51a7cf8948a3/include/rmm/cuda_stream_view.hpp#L58

```suggestion
DEFAULT_STREAM = Stream._from_cudaStream_t(rmm::cuda_stream_default().value())
```

harrism replied:
I don't know if that's correct Cython syntax. But you get the idea.

shwina replied:
Thanks! That worked.

kkraus14 commented on python/rmm/_cuda/stream.pyx:
```diff
@@ -0,0 +1,120 @@
+            pass
+        raise TypeError(f"Cannot create stream from {type(obj)}")
+
+    cdef void _init_with_new_cuda_stream(self) except *:
```
Should this be `nogil`?

shwina replied:
It can't because we're doing Pythony things inside

kkraus14 commented on python/rmm/_lib/cuda_stream.pxd:
```diff
@@ -0,0 +1,38 @@
+        void synchronize_no_throw()
+
+
+@cython.final
+cdef class CudaStream:
```
I know we used `cython.final` to allow making `is_valid` `nogil`, but this would prevent someone from inheriting `CudaStream`. Is this a reasonably fair restriction or would this become prohibitive?

shwina replied:
I think it's fine until it's not :) 

Although I suspect inheriting a CudaStream isn't something people will likely want to do.

harrism commented:
> > cudaStreamDefault is an option to cudaStreamCreateWithFlags!
> 
> Wow, I definitely missed this. Thanks! So does the default constructor of `rmm::cuda_stream_view()` then provide a view into `(cudaStream_t)0x0`?

The whole world missed it. Just do a github search for `cudaStreamDefault` and you'll find it used as a stream handle in many projects.  It was a design oversight, most likely.

harrism commented on python/rmm/_cuda/stream.pyx:
```diff
@@ -0,0 +1,117 @@
+
+
+DEFAULT_STREAM = Stream._from_cudaStream_t(<cudaStream_t>0)
+LEGACY_DEFAULT_STREAM = Stream._from_cudaStream_t(cudaStreamLegacy)
+PER_THREAD_DEFAULT_STREAM = Stream._from_cudaStream_t(cudaStreamPerThread)
```
And for consistency, you would change these:

```suggestion
LEGACY_DEFAULT_STREAM = Stream._from_cudaStream_t(rmm::cuda_stream_legacy().value())
PER_THREAD_DEFAULT_STREAM = Stream._from_cudaStream_t(rmm::cuda_stream_per_thread().value())
```

shwina replied:
Changed to use RMMs definitions

harrism commented on python/rmm/_lib/lib.pxd:
```diff
@@ -24,6 +24,9 @@ ctypedef pair[const char*, unsigned int] caller_pair
 cdef extern from * nogil:
 
     ctypedef void* cudaStream_t "cudaStream_t"
+    cdef cudaStream_t cudaStreamDefault
+    cdef cudaStream_t cudaStreamLegacy
+    cdef cudaStream_t cudaStreamPerThread
```
And then you could get rid of these.

shwina replied:
Done

harrism reviewed (changes requested):
Suggest using RMM's convenience cuda_stream_view wrappers instead of literal zero and/or lower-level handles.

harrism commented on python/rmm/_lib/cuda_stream_view.pxd:
```diff
@@ -20,8 +20,13 @@ cdef extern from "rmm/cuda_stream_view.hpp" namespace "rmm" nogil:
 
     cdef bool operator==(cuda_stream_view const, cuda_stream_view const)
+
+    const cuda_stream_view cuda_stream_default
```
So weird that these don't have to be declared as functions!

shwina replied:
Why would they be? Aren't these constexpr variables in C++?

harrism replied:
I'm sorry, I was thinking they were `constexpr` functions in C++. But they are variables, not functions. But note only one of them is `constexpr` (because you can't have a C-style cast (`reinterpret_cast`) in a `constexpr` expression, and CUDA defines `cudaStreamPerThread` using a c-style cast.

kkraus14 commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 - PR #596 Add `tracking_memory_resource_adaptor` to help catch memory leaks
 - PR #608 Add stream wrapper type
 - PR #632 Add RMM Python docs
+- PR #636 Add a `Stream` class to manage CUDA streams in Python
```
this needs to be moved to 0.18
